src/gateway.c

```diff
@@ -316,10 +318,35 @@ static void prepareBarrierCb(struct barrier *barrier, int status)
 		return;
 	}
 
-	response.db_id = (uint32_t)req->db_id;
-	response.id = (uint32_t)stmt->id;
-	response.params = (uint64_t)sqlite3_bind_parameter_count(stmt->stmt);
-	SUCCESS(stmt, STMT);
+	if (req->schema == 0 && *tail != '\0') {
+		stmt__registry_del(&g->stmts, stmt);
+		failure(req, SQLITE_ERROR, "nonempty statement tail");
+		return;
+	}
+
+	switch (req->schema) {
+		case 0:
+			response.db_id = (uint32_t)req->db_id;
+			response.id = (uint32_t)stmt->id;
+			response.params = (uint64_t)sqlite3_bind_parameter_count(stmt->stmt);
+			SUCCESS_V0(stmt, STMT);
+			break;
+		case 1:
+			response_with_offset.db_id = (uint32_t)req->db_id;
+			response_with_offset.id = (uint32_t)stmt->id;
+			response_with_offset.params = (uint64_t)sqlite3_bind_parameter_count(stmt->stmt);
+			response_with_offset.offset = (uint64_t)(tail - sql);
+
+			n = response_stmt_with_offset__sizeof(&response_with_offset);
```

Comment:
Is it not possible to use a `SUCCES_V1`-like macro here?

Reply:
If you can create some macro that's reusable, we can use it for that too, yes. Now there's some code duplication, and it just feels it could be handled somewhat cleaner.